langchain/vectorstores/matching_engine.py

```diff
@@ -0,0 +1,373 @@
+        )
+    
+    def _create_endpoint_by_id(
+        self,
+        endpoint_id: str
+    ) -> "aiplatform.MatchingEngineIndexEndpoint":
+        """Creates a MatchingEngineIndexEndpoint object by id.
+
+        Args:
+            endpoint_id: The created endpoint id. TODO(scafati98) link notebook
+
+        Returns:
+            A configured MatchingEngineIndexEndpoint.
+        """
+        logger.debug(f"Creating endpoint with id {endpoint_id}.")
+        return aiplatform.MatchingEngineIndexEndpoint(
+            index_endpoint_name=endpoint_id,
+            project=self.project_id,
+            location=self.region,
+            credentials=self.credentials,
+        )
+
+    def add_texts(
+        self,
+        texts: Iterable[str],
+        metadatas: Optional[List[dict]] = None,
+        **kwargs: Any,
+    ) -> List[str]:
+        """Run more texts through the embeddings and add to the vectorstore.
+
+        Args:
+            texts: Iterable of strings to add to the vectorstore.
+            metadatas: Optional list of metadatas associated with the texts.
+            kwargs: vectorstore specific parameters.
+
+        Returns:
+            List of ids from adding the texts into the vectorstore.
+        """
+        logger.debug(f"Embedding documents.")
+        embeddings = self.embedder.embed_documents(list(texts))
+        jsons = []
+        ids = []
+        # Could be improved with async.
+        for embedding, text in zip(embeddings, texts):
+            id = str(uuid.uuid4())
+            ids.append(id)
+            jsons.append({
+                "id": id,
+                "embedding": embedding
+            })
+            self._upload_to_gcs(text, f"documents/{id}")
+
+        logger.debug(f"Uploaded {len(ids)} documents to GCS.")
+
+        # Creating json lines from the embedded documents.
+        result_str = "\n".join([json.dumps(x) for x in jsons])
+
+        filename_prefix = f"indexes/{uuid.uuid4()}"
+        filename = f"{filename_prefix}/{time.time()}.json"
+        self._upload_to_gcs(result_str, filename)
+        logger.debug(f"Uploaded updated json with embeddings to "
+                     f"{self.gcs_bucket_uri}/{filename}.")
+
+        self.index = self.index.update_embeddings(
+            contents_delta_uri=f"gs://{self.gcs_bucket_uri}/{filename_prefix}/"
+        )
+
+        logger.debug(f"Updated index with new configuration.")
+
+        return ids
+    
+    def _upload_to_gcs(self, data: str, gcs_location: str) -> None:
+        """Uploads data to gcs_location.
+
+        Args:
+            data: The data that will be stored.
+            gcs_location: The location where the data will be stored.
+        """
+        client = self._get_gcs_client()
+        bucket = client.get_bucket(self.gcs_bucket_uri)
+        blob = bucket.blob(gcs_location)
+        blob.upload_from_string(data)
+
+    def _get_gcs_client(self) -> storage.Client:
+        """Lazily creates a GCS client.
```

Comment:
are there major benefits to lazy creation?

Reply:
The benefit is that if the user does not use the `add_texts` method the client would never be created. I wouldn't call it a *major* benefit. @dev2049

Reply:
I think it's better to require teh client in the class initialization, which also let's us collapse the credentials and project_id since they're contained within the client.